borgmatic/hooks/postgresql.py

```diff
@@ -22,14 +22,18 @@ def make_dump_path(location_config):  # pragma: no cover
         location_config.get('borgmatic_source_directory'), 'postgresql_databases'
     )
 
-
-def make_extra_environment(database):
+def make_extra_environment(database, restore_connection_params=None):
     '''
     Make the extra_environment dict from the given database configuration.
+    If restore connection params are given, this is for a restore operation.
     '''
     extra = dict()
-    if 'password' in database:
-        extra['PGPASSWORD'] = database['password']
+
+    try:
+        extra['PGPASSWORD'] = restore_connection_params.get('password') or database['restore_password'] or database['password']
```

Comment:
I _think_ `database['restore_password']` needs to be `database.get('restore_password')` here, or you'll never get to `database['password']`.

Also, I'm pretty sure using the `'restore_password'` value from the `database` configuration needs to gate on whether `restore_connection_params` is non-`None`—since that's now your "flag" for whether this is a restore, right?

Soapbox time: Sometimes when I'm working on gnarly logic like this, I actually find it easier to write the tests _as_ I'm developing the function. That way, I can convince myself that the code I'm writing satisfies each use case. Doesn't always work (sometimes I still screw it up even after writing tests!), but sometimes it actually helps.